python/node_similarity.py

```diff
@@ -0,0 +1,285 @@
+    :param node2: The second node
+    :type node2: mgp.Vertex
+
+    :return: The overlap similarity of the first and the second node
+    :rtype: float
+
+    """
+
+    overlap_similarity: float
+
+    neighbors1: set
+    neighbors2: set
+
+    if node1 == node2:
+        return 1.0
+    else:
+        overlap_similarity = 0.0
+
+    neighbors1 = _get_neighbors(node1)
+    neighbors2 = _get_neighbors(node2)
+
+    denominator = min(len(neighbors1), len(neighbors2))
+
+    if denominator != 0:
+        overlap_similarity = len(neighbors1 & neighbors2) / denominator
+
+    return overlap_similarity
+
+
+def _calculate_cosine(node1: mgp.Vertex, node2: mgp.Vertex) -> float:
+    """This method calculates the cosine similarity between two nodes.
+
+    :param node1: The first node
+    :type node1: mgp.Vertex
+    :param node2: The second node
+    :type node2: mgp.Vertex
+
+    :return: The cosine similarity of the first and the second node
+    :rtype: float
+    """
+
+    cosine_similarity: float
+
+    neighbors1: set
+    neighbors2: set
+
+    if node1 == node2:
+        return 1.0
+    else:
+        cosine_similarity = 0.0
+
+    neighbors1 = _get_neighbors(node1)
+    neighbors2 = _get_neighbors(node2)
+
+    denominator = sqrt(len(neighbors1) * len(neighbors2))
+
+    if denominator != 0:
+        cosine_similarity = len(neighbors1 & neighbors2) / denominator
+
+    return cosine_similarity
+
+
+def _calculate_similarity(
+    node1: Union[mgp.Vertex, Tuple[mgp.Vertex]],
+    node2: Union[mgp.Vertex, Tuple[mgp.Vertex]],
+    method: Callable,
+    mode: str,
+) -> List[Tuple[mgp.Vertex, mgp.Vertex, float]]:
+    """This method calculates the similarity of nodes for given method and mode.
+
+    :param node1: The first node or tuple of nodes
+    :type node1: mgp.Vertex, tuple
+    :param node2: The second node or tuple of nodes
+    :type node2: mgp.Vertex, tuple
+    :param method: Similarity measure which will be used for calculating the similarity between nodes.
+    Currently available are `_calculate_jaccard`, `_calculate_overlap` and `_calculate_cosine`
+    :type method: function
+    :param mode: Can be `p` (or `pairwise`) for pairwise similarity calculation or `c` (or `cartesian`) for
+    calculating the similarity of the Cartesian product of given tuples
+    :type mode: str
+
+    :raises TypeError: Occurs if there's a type mismatch for passed arguments
+    :raises ValueError: Occurs if any passed argument is invalid
+
+    :return: Returns the calculated similarity between nodes in a list of mgp.Records. Each mgp.Record consist of
+    the first node, the second node and the similarity between them.
+    :rtype: list
+    """
+
+    # clear cache
+    neighbors_dict.clear()
+
+    nodes1: tuple
+    nodes2: tuple
+
+    if isinstance(node1, mgp.Vertex):
+        nodes1 = tuple([node1])
+    elif isinstance(node1, tuple):
+        nodes1 = node1
+    else:
+        raise TypeError("Invalid type of first argument.")
```

Comment:
Like my comment from below, I think this can go up (needs to be checked first)

Reply:
And again, I prefer the way I wrote it. I think it's easier to read.